core/src/main/java/de/bwaldvogel/mongo/backend/AbstractMongoBackend.java

```diff
@@ -236,7 +242,27 @@ public Document handleCommand(Channel channel, String databaseName, String comma
             return handleAdminCommand(command, query);
         } else {
             MongoDatabase db = resolveDatabase(databaseName);
-            return db.handleCommand(channel, command, query);
+            Document doc = db.handleCommand(channel, command, query);
```

Comment:
Why not decorate existing `MongoDatabase` with oplog functionality ? 

### Example
```java
/**
  * Intercepts all the calls and records modifications to a collection
 */
class OpLogDatabase implements MongoDatabase {
   private final MongoDatabase delegate;
   private final MongoCollection oplog;
   
   public OpLogDatabase(MongoDatabase delegate) {
      this.delegate = delegate;
	  this.oplog = delegate.getCollection("local.oplog.rs");
   }
   
   public Document handleCommand(Channel channel, String command, Document query) {     
	 
	 Docoument doc = delegate.handleCommand(channel, command, query);
     if ("insert".equalsIgnoreCase(command) || "update".equialsIgnoreCase(command)) {
	    handleOplog(channel, command, query, doc);
	 }	 
   }  

   private void handleOplog(Channel channel, String command, Document query, String databaseName) {
     // ....
   }   
}
```

Avoid adding too many custom `if`s when you can achieve same functionality using composition.

Reply:
Following up on our offline conversation on the idea to compose MongoBackend instead of MongoDatabase. 
I'm concerned it will lead to changes on the public api, for instance
`server = new MongoServer(new MemoryBackend());`
will be
`server = new MongoServer(new OplogBackend(new MemoryBackend()));`
or
`server = new MongoServer(new OplogMemoryBackend());`

I think we should avoid changing the public api and make the oplog functionality a default for all backends, favouring inheritance over composition in this particular case

`public abstract class OplogBackend extends MongoAbstractBackend {
...
}
`

`public class MemoryBackend extends OplogBackend {
...
}
`